Fix various namespace prefix conflict resolution bugs

nielsdos opened the pull request:
The output could still be improved by removing redundant namespace
declarations, but that's another issue. At least the output is
correct now.

Depends on the fix in GH-11776, so the first commit just applies that here.
Will clean that up once that PR is merged of course.

Girgias reviewed (approved):
Looks reasonable to me, don't know if @theseer wants to double-check the test output.

theseer commented on ext/dom/tests/DOMDocument_createAttributeNS_prefix_conflict.phpt:
```diff
@@ -0,0 +1,82 @@
+string(18) "http://php.net/ns2"
+string(18) "http://php.net/ns3"
+<?xml version="1.0"?>
+<container xmlns="http://php.net/ns1" xmlns:default="http://php.net/ns2" xmlns:default1="http://php.net/ns3" xmlns:default2="http://php.net/ns4" default2:hello=""/>
```
For the record, this is what firefox does (formatting by me):

```xml
'<container xmlns="http://www.w3.org/1999/xhtml" 
   a0:hello="" xmlns:a0="https://php.net/ns1"
   a1:hello="" xmlns:a1="https://php.net/ns2"
   a2:hello="" xmlns:a2="https://php.net/ns3"
   a3:hello="" xmlns:a3="https://php.net/ns4">
</container>'
```

nielsdos replied:
First of all thanks for checking!

> This test is confusing me, as it claims to test createAttributeNS but it mainly tests setAttributeNodeNS.

All subtests test the behaviour of `createAttributeNS`, it was there that there was an implementation bug so that's why I named the test like this. The first half tests the behaviour with `setAttributeNode` (no NS) and the second half `setAttributeNodeNS`.

> The W3C-Spec (https://www.w3.org/TR/DOM-Level-2-Core/core.html) for setAttributeNodeNS says: "If an attribute with that local name and that namespace URI is already present in the element, it is replaced by the new one." This is not the case for your example, as the NS differes, so it should return null for all cases. As the NS is different, I'd expect all 4 attributes to be set.

Note that the first half tests setAttributeNode, not setAttributeNodeNS: https://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/core.html#ID-887236154
However I need to recheck if I did the nodeName test correctly for all cases, I'll do that tomorrow.

> The <container>-Node was created without a namespace. Now, when re-parsing this, the node would be in https://php.net/ns1. That seems wrong.

Good catch! Unfortunately it looks like this has been a bug for a long time :(: https://3v4l.org/kid5o
Fixing it in this PR is appropriate, so I'll do that.

> If your implementation were correct, the replaced attribute's namespace should be dropped, because now they are redundant.

I don't think this is required by spec right?

theseer replied:
> 
> > The W3C-Spec (https://www.w3.org/TR/DOM-Level-2-Core/core.html) for setAttributeNodeNS says: "If an attribute with that local name and that namespace URI is already present in the element, it is replaced by the new one." This is not the case for your example, as the NS differes, so it should return null for all cases. As the NS is different, I'd expect all 4 attributes to be set.
> 
> Note that the first half tests setAttributeNode, not setAttributeNodeNS: https://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/core.html#ID-887236154 However I need to recheck if I did the nodeName test correctly for all cases, I'll do that tomorrow.

I did mention that these longish tests with the expects way later are confusing ;)  You're of course right, I missed the NS not being used when scrolling back. I still believe mixing DOM1 and DOM2+ APIs should just be forbidden ;) I know that's not according to the Spec though...

Then I retract my previous statement regarding the `setAttributeNode` behavior: The result makes sense, given the nodeName is identical and the non-NS-aware API should probably just ignore the Namespace.

Which would lead to the question, whether those xmlns-declarations should be there at all: _To add a new attribute node with a qualified name and namespace URI, use the setAttributeNodeNS method._  Wouldn't that imply they should be ignored?

> 
> > If your implementation were correct, the replaced attribute's namespace should be dropped, because now they are redundant.
> 
> I don't think this is required by spec right?

Not as far as i can tell. Leaving things laying around is still a bit messy ;)

nielsdos replied:
> I did mention that these longish tests with the expects way later are confusing ;)

Sorry, you're right. I'm going to split this test into two files, I should've done so from the start.

> I still believe mixing DOM1 and DOM2+ APIs should just be forbidden ;) I know that's not according to the Spec though...

I agree, it's ugly.

> Which would lead to the question, whether those xmlns-declarations should be there at all: To add a new attribute node with a qualified name and namespace URI, use the setAttributeNodeNS method. Wouldn't that imply they should be ignored?

HMM. Good question. The fact that this so vaguely worded certainly doesn't help... I'm going to check in detail tomorrow.

> Not as far as i can tell. Leaving things laying around is still a bit messy ;)

True. I'm going to fix this separately in a follow-up, as this is not the only place where it gets messy. I want to get the output correct first :)

theseer reviewed (commented):
Unfortunately, I believe to have found some issues...

theseer commented on ext/dom/tests/DOMDocument_createAttributeNS_prefix_conflict.phpt:
```diff
@@ -0,0 +1,82 @@
+string(18) "http://php.net/ns2"
+string(18) "http://php.net/ns3"
+<?xml version="1.0"?>
+<container xmlns="http://php.net/ns1" xmlns:default="http://php.net/ns2" xmlns:default1="http://php.net/ns3" xmlns:default2="http://php.net/ns4" xmlns:foo="http://php.net/ns1" default2:hello=""/>
```
Same problem here.

theseer replied:
Same problem as above.

theseer commented on ext/dom/tests/DOMDocument_createAttributeNS_prefix_conflict.phpt:
```diff
@@ -0,0 +1,82 @@
+NULL
+string(18) "http://php.net/ns1"
+<?xml version="1.0"?>
+<container xmlns:foo="http://php.net/ns1" foo:hello=""/>
```
Same problem as above.

theseer commented on ext/dom/tests/DOMDocument_importNode_attribute_prefix_conflict.phpt:
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0"?>
+<container xmlns:foo="http://php.net" foo:bar="yes"/>
+<?xml version="1.0"?>
+<container xmlns="http://php.net" xmlns:foo="http://php.net/2" xmlns:default="http://php.net" default:bar="yes"/>
```
While technically correct, shouldn't/couldn't the `default:`-prefix be avoided, given `container` is in the same namespace?

theseer replied:
This even feels counter intuitive, given the next test actually even ignores the namespace.

nielsdos replied:
Default namespaces don't apply to attributes, so we need the explicit default: prefix.

nielsdos replied:
That's also why the next test ignores the default namespace.

nielsdos replied:
Sorry, I don't understand what you mean.

If we write `bar` instead of `default:bar`, then the bar attribute belongs to no namespace. Example from Javascript:
```js
dom = (new DOMParser).parseFromString('<?xml version="1.0"?><container xmlns="http://php.net" bar="yes"/>', 'text/xml')
dom.documentElement.attributes[1].namespaceURI // outputs: null
```

versus:
```js
dom = (new DOMParser).parseFromString('<?xml version="1.0"?><container xmlns="http://php.net" xmlns:default="http://php.net" default:bar="yes"/>', 'text/xml')
dom.documentElement.attributes[2].namespaceURI // outputs: http://php.net
```

theseer replied:
You're right. Disregard. I should probably go sleep :)

theseer commented:
For the record: The rest looks good to me :)

nielsdos commented:
I've fixed the default namespace issue and split the tests into different files.

> Which would lead to the question, whether those xmlns-declarations should be there at all: To add a new attribute node with a qualified name and namespace URI, use the setAttributeNodeNS method. Wouldn't that imply they should be ignored?

I've thought about this, and I've done a little digging into the history of this method. Up until 2015, this method used with namespaced attributes had different behaviours in different browsers. WHATWG then decided to alias setAttributeNode to setAttributeNodeNS. I can't make this change here right now because of BC issues, but I plan to address this in an upcoming effort (in PHP 8.4) for opt-in spec compliance.

I've also retargeted this to master, because of the behavioural differences and I don't want to get complaints about this. If necessary, I can backport the default namespace handling without the other behaviour changes in this PR.

Finally, since the attribute handling influences how namespaces can be reused, I had to remove the namespace reuse code from `dom_get_ns`, this caused me to run into https://bugs.php.net/bug.php?id=80927 for one of the tests, so I modified the test for now until I figure out a way to fix that bug.

nielsdos commented:
> (...) this caused me to run into https://bugs.php.net/bug.php?id=80927 for one of the tests, so I modified the test for now until I figure out a way to fix that bug.

Filed https://github.com/php/php-src/pull/11892 as a draft PR to deal with this
